File: C++/nextPermutation.cpp
// Time Complexity: O(n)
// Space Complexity: O(1)

class Solution {
    public:
        void nextPermutation(vector<int> &num) {
            nextPermutation(begin(num), end(num));
        }

    private:
        template<typename BidiIt>
            bool nextPermutation(BidiIt begin, BidiIt end) {
                const auto rbegin = reverse_iterator<BidiIt>(end);
                const auto rend = reverse_iterator<BidiIt>(begin);

                // find the firt element (pivot) which is less than its successor
                auto pivot = next(rbegin);
                while(pivot != rend && *pivot >= *prev(pivot)) {
                    ++pivot;
                }

                // no next permutation, just reverse the whole sequence
                if(pivot == rend) {
                    reverse(rbegin, rend);
                    return false;
                }

                // find the number which is greater than pivot, and swap it with pivot
                auto change = find_if(rbegin, pivot, bind1st(less<int>(), *pivot));
                swap(*change, *pivot);

                // make the sequence after pivot non-descending
                reverse(rbegin, pivot);

                return true; // return next permutation
            }
};

File: Python/next-permutation.py
# Time:  O(n)
# Space: O(1)

class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        k, l = -1, 0
        for i in reversed(xrange(len(nums)-1)):
            if nums[i] < nums[i+1]:
                k = i
                break
        else:
            nums.reverse()
            return

        for i in reversed(xrange(k+1, len(nums))):
            if nums[i] > nums[k]:
                l = i
                break
        nums[k], nums[l] = nums[l], nums[k]
        nums[k+1:] = nums[:k:-1]
        

# Time:  O(n)
# Space: O(1)
class Solution2(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        k, l = -1, 0
        for i in xrange(len(nums)-1):
            if nums[i] < nums[i+1]:
                k = i

        if k == -1:
            nums.reverse()
            return

        for i in xrange(k+1, len(nums)):
            if nums[i] > nums[k]:
                l = i
        nums[k], nums[l] = nums[l], nums[k]
        nums[k+1:] = nums[:k:-1]
